### ISSUE TITLE
How to add new lanuages 

### ISSUE THREAD

**Issue by hxue3** [facebookresearch/CodeGen#42]
I am wondering if I can add new lanuages for code translation, for example I want to translate COBOL code to python

Do you have any tips if you can explain briefly that what I should do

**Comment by raffian** [comment]
@brozi 

I need a processor for C. Should I follow the same steps?

**Comment by brozi** [comment]
Yes and you can probably adapt the C++ processor for C https://github.com/facebookresearch/CodeGen/blob/c83433217fdba964d1f15aa4d45a78c75d6bfa12/codegen_sources/preprocessing/lang_processors/cpp_processor.py

**Comment by raffian** [comment]
@brozi 

Thanks for the quick reply.

Would you be willing to share the queries (BigQuery) used to extract monolingual cpp data from the github public dataset? Whitepaper suggests I'll need to do that for C, but can probably reuse the java dataset under `data/test_dataset` as-is given java is our target, correct?

**Comment by arpieb** [comment]
@raffian @brozi did you ever suss out what the queries should look like?  Hoping to leverage this to train a Delphi <-> Python3 transcoder model.

Thanks!

**Comment by raffian** [comment]
@arpieb 

Unfortunately, no - I gave up on this approach.

I switched my focus to ANTLR4, it's mature, stable, and comes with lots of grammars for parsing nearly every programming language ever created, though surprisingly Delphi is missing from the list.
https://github.com/antlr/grammars-v4

I found this one - unofficial I guess so your mileage may vary. 
https://github.com/gotthardsen/Delphi-ANTRL4-Grammar/blob/master/Delphi.g4

If you go down the ANTLR4 path for code translation, take my advice and do the training at udemy, it's worth it.  Understanding ANTLR fundamentals are essential to using it effectively, otherwise, you'll just get frustrated with it.
https://www.udemy.com/course/antlr-programming-masterclass-with-python

Good luck,
raffian

**Comment by raffian** [comment]
@arpieb 

Best place to start with ANTLR: https://www.antlr.org/

**Comment by arpieb** [comment]
@raffian thanks for the links, will check them out!  really hoping to find something that can intelligently perform the translation, or at least a large part of it.
